### PR TITLE
{ARO} `az aro`: Bump API version from 2022-04-01 to 2022-09-04

### PR DESCRIPTION
**Pull request by dem4gus**
**Related command**

`az aro`

**Description**

* Bump API version from v2022-04-01 to v2022-09-04
* Add NetworkContributor role to NAT Gateway
* Change `--pull-secret` parameter to not require @ prefix on filename

**Testing Guide**

**History Notes**

{ARO}: Bump API version from 2022-04-01 to 2022-09-04
{ARO}: Refactor validators
[ARO] `az aro create/update`: Add NetworkContributor role to NAT Gateways in Cluster Resource Group when creating or updating clusters
[ARO] `az aro create`: Change `--pull-secret` parameter to no longer require `@` prefix on filenames

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

File: src/azure-cli/azure/cli/command_modules/aro/_params.py [hunk 1]
```diff
@@ -10,13 +10,15 @@
 from azure.cli.command_modules.aro._validators import validate_disk_encryption_set
 from azure.cli.command_modules.aro._validators import validate_domain
 from azure.cli.command_modules.aro._validators import validate_pull_secret
+from azure.cli.command_modules.aro._validators import validate_sdn
```
Comment (SudoBrendan) [inline src/azure-cli/azure/cli/command_modules/aro/_params.py:13]: I don't think SDN choice is in any GA API version, and is not planned to be a choice. I think everything related to SDN should be removed from this PR - I'll flag them.

File: src/azure-cli/azure/cli/command_modules/aro/_params.py [hunk 2]
```diff
@@ -54,12 +56,20 @@ def load_arguments(self, _):
...
         c.argument('pod_cidr',
                    help='CIDR of pod network. Must be a minimum of /18 or larger.',
                    validator=validate_cidr('pod_cidr'))
         c.argument('service_cidr',
                    help='CIDR of service network. Must be a minimum of /18 or larger.',
                    validator=validate_cidr('service_cidr'))
+        c.argument('software_defined_network', arg_type=get_enum_type(['OVNKubernetes', 'OpenShiftSDN']),
+                   options_list=['--software-defined-network-type', '--sdn-type'],
+                   help='SDN type either "OpenShiftSDN" (default) or "OVNKubernetes"',
+                   validator=validate_sdn)
```
Comment (SudoBrendan) [inline src/azure-cli/azure/cli/command_modules/aro/_params.py:72]: SDN - remove.

File: src/azure-cli/azure/cli/command_modules/aro/_validators.py [hunk 3]
```diff
@@ -99,13 +106,27 @@ def validate_pull_secret(namespace):
...
+
+
+def validate_sdn(namespace):
+    if namespace.software_defined_network is None:
+        return
+
+    target_values = ['OVNKubernetes', 'OpenshiftSDN']
+    if namespace.software_defined_network not in target_values:
+        raise InvalidArgumentValueError(
+            f"Invalid --software-defined-network '{namespace.software_defined_network}'.")
```
Comment (SudoBrendan) [inline src/azure-cli/azure/cli/command_modules/aro/_validators.py:129]: SDN - remove.

File: src/azure-cli/azure/cli/command_modules/aro/custom.py [hunk 4]
```diff
@@ -48,6 +47,7 @@ def aro_create(cmd,  # pylint: disable=too-many-locals
                client_secret=None,
                pod_cidr=None,
                service_cidr=None,
+               software_defined_network=None,
```
Comment (SudoBrendan) [inline src/azure-cli/azure/cli/command_modules/aro/custom.py:50]: SDN - remove.

File: src/azure-cli/azure/cli/command_modules/aro/custom.py [hunk 5]
```diff
@@ -110,6 +112,7 @@ def aro_create(cmd,  # pylint: disable=too-many-locals
         network_profile=openshiftcluster.NetworkProfile(
             pod_cidr=pod_cidr or '10.128.0.0/14',
             service_cidr=service_cidr or '172.30.0.0/16',
+            software_defined_network=software_defined_network or 'OpenShiftSDN',
```
Comment (SudoBrendan) [inline src/azure-cli/azure/cli/command_modules/aro/custom.py:115]: SDN - remove.

**Review by SudoBrendan — CHANGES_REQUESTED** [review]
Other than holding for multi-version to be GA globally, I commented on some code that relates to cluster SDN choice, which I want to confirm is actually part of the API version we are using - I don't think it is, since we have historically taken the stance that we only support 1 type for specific OpenShift versions.

**Comment by yonzhan** [comment]
ARO

File: src/azure-cli/azure/cli/command_modules/aro/custom.py (hunk 4 above)
Comment (dem4gus) [reply src/azure-cli/azure/cli/command_modules/aro/custom.py]: Removed

File: src/azure-cli/azure/cli/command_modules/aro/_params.py (hunk 2 above)
Comment (dem4gus) [reply src/azure-cli/azure/cli/command_modules/aro/_params.py]: Removed

File: src/azure-cli/azure/cli/command_modules/aro/_validators.py (hunk 3 above)
Comment (dem4gus) [reply src/azure-cli/azure/cli/command_modules/aro/_validators.py]: Removed

File: src/azure-cli/azure/cli/command_modules/aro/custom.py (hunk 5 above)
Comment (dem4gus) [reply src/azure-cli/azure/cli/command_modules/aro/custom.py]: Removed

File: src/azure-cli/azure/cli/command_modules/aro/_params.py (hunk 1 above)
Comment (dem4gus) [reply src/azure-cli/azure/cli/command_modules/aro/_params.py]: I think I got them all, thanks for the extra set of eyes.

**Comment by dem4gus** [comment]
The CI pipeline is failing with the following error. I'm not sure what I can do to fix this, since it seems the extension that is failing is wholly unrelated to `az aro`

```
Verifying extension: healthcareapis
WARNING: Unable to load extension 'alias: cannot import name 'soft_unicode' from 'markupsafe' (/opt/az/azcliextensions/alias/markupsafe/__init__.py)'. Use --debug for more information.
WARNING: This package of the Batch Extensions module supports Azure Batch up to version 12.2. The current version 13.0.0 has not been tested for compatibility.
WARNING: This package of the Batch Extensions module supports Azure Batch Management up to version 16.3. The current version 17.0.0 has not been tested for compatibility.
ERROR: The checksum of the extension does not match the expected value. Use --debug for more information.
```

File: src/azure-cli/azure/cli/command_modules/aro/_client_factory.py [hunk 6]
```diff
@@ -10,6 +10,6 @@
 
 def cf_aro(cli_ctx, *_):
     client = get_mgmt_service_client(
-        cli_ctx, AzureRedHatOpenShiftClient).open_shift_clusters
+        cli_ctx, AzureRedHatOpenShiftClient)
 
     return client
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/aro/_client_factory.py:15]: Usually, we will add another client_factory function to return open_shift_clusters. Like this https://github.com/Azure/azure-cli/blob/ba12dedfb2f60988cb78d4b23103412eae45bd96/src/azure-cli/azure/cli/command_modules/cdn/_client_factory.py#L7-L14

File: src/azure-cli/setup.py [hunk 7]
```diff
@@ -111,7 +111,7 @@
     'azure-mgmt-rdbms~=10.2.0b6',
     'azure-mgmt-recoveryservicesbackup~=5.1.0b1',
     'azure-mgmt-recoveryservices~=2.1.0',
-    'azure-mgmt-redhatopenshift==1.1.0',
+    'azure-mgmt-redhatopenshift==1.2.0',
```
Comment (kairu-ms) [inline src/azure-cli/setup.py:114]: Usually, we will use `~=1.2.0` instead of `==1.2.0` in `setup.py` file for SDK version.

Comment (dem4gus) [reply src/azure-cli/setup.py]: This has been changed.

File: src/azure-cli/azure/cli/command_modules/aro/_client_factory.py (hunk 6 above)
Comment (dem4gus) [reply src/azure-cli/azure/cli/command_modules/aro/_client_factory.py:15]: I will create a work item for us to address both this and a client factory for the newly created `aro_get_versions` command.

**Review by kairu-ms — APPROVED** [review]

**Pull request merged**